### PR TITLE
Move Cirq-FT infrastructure to Qualtran

### PR DESCRIPTION
**Pull request by tanujkhattar**
This PR is the first big PR that moves code from Cirq-FT to Qualtran as part of [Qualtran & Cirq-FT Integration](https://docs.google.com/document/d/1miypjcCNt6JV_qAypUVPy5eH3EmKadLqeZomA5QoWK0/edit?usp=sharing&resourcekey=0-ciQ8jPfQmK8gtey2lNKesw)

Some key features:
- `GateWithRegisters` now exists in `Qualtran` and is both a `cirq.Gate` and a `Bloq`.
- As a proof of concept, `And` and `MultiAnd` bloq's in `Qualtran` now derive from the `GateWithRegisters` class. They pass all existing tests and are used in cirq-ft style tests as well. 
- `CirqGateAsBloq` and `BloqAsCirqGate` don't depend upon `cirq_ft` at the top-level anymore. There still exists some temporary glue code that needs to support both `cirq_ft.GateWithRegisters` and `qualtran.GateWithRegisters` as part of the transition period, but that'll be gone once the migration is completed. 
- Other cirq infra code, like `TComplexity`, `decompose_protocol` etc. also make their way into qualtran as part of the `cirq_interop` repository. In follow-up PRs, we can rearrange this code (for example: move all or parts of `TComplexity` to `resource_counting/` instead of `cirq_interop`). 

cc @fdmalone @mpharrigan

### CODE REVIEW

File: qualtran/cirq_interop/decompose_protocol.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+from typing import Any, FrozenSet, Sequence
+
+import cirq
+from cirq.protocols.decompose_protocol import DecomposeResult
+
+_FREDKIN_GATESET = cirq.Gateset(cirq.FREDKIN, unroll_circuit_op=False)
+
+
+def _fredkin(qubits: Sequence[cirq.Qid], context: cirq.DecompositionContext) -> cirq.OP_TREE:
```
Comment (mpharrigan) [inline qualtran/cirq_interop/decompose_protocol.py:23]: reminder that per our agreement if this code still exists like 1 month from now I will unceremoniously remove it

**Review by mpharrigan — COMMENTED** [review]
Can you comment on lingering compatibility issues? It looks like `Signature` is almost the same but not quite

File: qualtran/bloqs/swap_network_cirq_test.py [hunk 2]
```diff
@@ -74,7 +74,7 @@ def test_swap_with_zero_gate(selection_bitsize, target_bitsize, n_target_registe
 def test_swap_with_zero_gate_diagram():
     gate = SwapWithZeroGate(3, 2, 4)
     q = cirq.LineQubit.range(cirq.num_qubits(gate))
-    circuit = cirq.Circuit(gate.on_registers(**cirq_ft.infra.split_qubits(gate.signature, q)))
+    circuit = cirq.Circuit(gate.on_registers(**split_qubits(gate.signature, q)))
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/bloqs/swap_network_cirq_test.py:77]: Both of these would be removed and replaced with `swap_network.py` from Cirq-FT which already derives from `GateWithRegisters` and thus would be a Bloq.

File: qualtran/cirq_interop/decompose_protocol.py (hunk 1 above)
Comment (tanujkhattar) [reply qualtran/cirq_interop/decompose_protocol.py:23]: I don't want to keep this code either. The project that takes care of supporting different decompositions when counting resources should remove this code as a first proof of concept. Whoever removes this code should make sure the T-counts of higher level primitives remain the same, and then I'm happy.

File: qualtran/cirq_interop/_bloq_to_cirq.py [hunk 3]
```diff
@@ -271,25 +268,26 @@ def _construct_op_from_gate(
     """
     all_quregs: Dict[str, 'CirqQuregT'] = {}
     out_quregs: Dict[str, 'CirqQuregT'] = {}
+    # def _cmp()
     for reg in gate.signature:
         full_shape = reg.shape + (reg.bitsize,)
-        if reg.side & cirq_ft.infra.Side.LEFT:
+        if Side(reg.side.value) & Side.LEFT:
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/cirq_interop/_bloq_to_cirq.py:274]: Added a TODO indicating that this is a short-lived code.

File: qualtran/bloqs/arithmetic_test.py [hunk 4]
```diff
@@ -207,3 +207,4 @@ def test_to_contiguous_index():
 
 def test_notebook():
     execute_notebook('arithmetic')
+    execute_notebook('comparison_gates')
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/bloqs/arithmetic_test.py:210]: Created two separate methods.

File: qualtran/_infra/registers.py [hunk 5]
```diff
@@ -168,13 +170,9 @@ def __len__(self) -> int:
 
     def get_cirq_quregs(self) -> Dict[str, 'NDArray[cirq.Qid]']:
         """Get arrays of cirq qubits for these registers."""
-        import cirq_ft
+        from qualtran._infra.gate_with_registers import get_named_qubits
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/_infra/registers.py:173]: Added a TODO indicating that this is a short-lived code.

File: qualtran/_infra/registers.py [hunk 6]
```diff
@@ -19,7 +19,7 @@
 from typing import Dict, Iterable, Iterator, List, overload, Tuple, TYPE_CHECKING
 
 import numpy as np
-from attr import frozen
+from attr import field, frozen
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/_infra/registers.py:22]: Changed

File: qualtran/cirq_interop/_cirq_to_bloq.py [hunk 7]
```diff
@@ -56,33 +57,21 @@ def short_name(self) -> str:
...
     def signature(self) -> 'Signature':
-        return Signature(
-            [
-                Register(reg.name, bitsize=reg.bitsize, shape=reg.shape, side=Side(reg.side.value))
-                for reg in self.cirq_registers
-            ]
-        )
+        if isinstance(self.gate, Bloq):
+            return self.gate.signature
+        import cirq_ft
```
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/cirq_interop/_cirq_to_bloq.py:62]: Yes, Added a TODO indicating that this is a short-lived code.

**Comment by tanujkhattar** [comment]
>  It looks like Signature is almost the same but not quite

Cirq-FT signature is lacking a few accessor methods (like `signature.lefts()` / `signature.rights()`) but it's a strict subset of `Qualtran.Signature`. Apart from that there should no difference. Is there a specific compatibility issue that you are concerned about?

**Comment by tanujkhattar** [comment]
@mpharrigan I've addressed all your comments. This is ready for another look.

File: qualtran/bloqs/swap_network_cirq_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mpharrigan) [reply qualtran/bloqs/swap_network_cirq_test.py:77]: are you going to do this?

[the remark this answers is not in the document]
Comment (tanujkhattar) [reply qualtran/bloqs/swap_network_cirq_test.py:77]: Yup

**Comment by mpharrigan** [comment]
> Apart from that there should no difference. Is there a specific compatibility issue that you are concerned about?

The `Side` enum thing is a difference I wouldn't have expected, a priori.

**Review by mpharrigan — APPROVED** [review]
sweeeeeeeeet

**Pull request merged**